java/adapter/avro/src/main/java/org/apache/arrow/AvroToArrowVectorIterator.java

```diff
@@ -0,0 +1,170 @@
+      Decoder decoder,
+      Schema schema,
+      BufferAllocator allocator,
+      int targetBatchSize) {
+
+    this.decoder = decoder;
+    this.schema = schema;
+    this.allocator = allocator;
+    this.targetBatchSize = targetBatchSize;
+
+  }
+
+  /**
+   * Create a ArrowVectorIterator to partially convert data.
+   */
+  public static AvroToArrowVectorIterator create(
+      Decoder decoder,
+      Schema schema,
+      BufferAllocator allocator,
+      int targetBatchSize) {
+
+    AvroToArrowVectorIterator iterator = new AvroToArrowVectorIterator(decoder, schema, allocator, targetBatchSize);
+    try {
+      iterator.initialize();
+      return iterator;
+    } catch (Exception e) {
+      iterator.close();
+      throw new RuntimeException("Error occurs while creating iterator.", e);
+    }
+  }
+
+  private void initialize() {
+    // create consumers
+    compositeConsumer = AvroToArrowUtils.createCompositeConsumer(schema, allocator);
+    List<FieldVector> vectors = new ArrayList<>();
+    compositeConsumer.getConsumers().forEach(c -> vectors.add(c.getVector()));
+    List<Field> fields = vectors.stream().map(t -> t.getField()).collect(Collectors.toList());
+    VectorSchemaRoot root = new VectorSchemaRoot(fields, vectors, 0);
+    rootSchema = root.getSchema();
+
+    load(root);
+  }
+
+  private void consumeData(VectorSchemaRoot root) {
+    int readRowCount = 0;
+    try {
+      while ((targetBatchSize == NO_LIMIT_BATCH_SIZE || readRowCount < targetBatchSize)) {
+        compositeConsumer.consume(decoder, root);
+        readRowCount++;
+      }
+      root.setRowCount(readRowCount);
+    } catch (EOFException eof) {
+      // reach the end of encoder stream.
+      root.setRowCount(readRowCount);
+    } catch (Exception e) {
+      compositeConsumer.close();
+      throw new RuntimeException("Error occurs while consuming data.", e);
+    }
+  }
+
+  // Loads the next schema root or null if no more rows are available.
+  private void load(VectorSchemaRoot root) {
+
+    Preconditions.checkArgument(root.getFieldVectors().size() == compositeConsumer.getConsumers().size(),
+        "Schema root vectors size not equals to consumers size.");
+
+    for (int i = 0; i < root.getFieldVectors().size(); i++) {
+      compositeConsumer.getConsumers().get(i).resetValueVector(root.getFieldVectors().get(i));
+    }
+
+    // consume data
+    consumeData(root);
+
+    if (root.getRowCount() == 0) {
+      root.close();
+      nextBatch = null;
+    } else {
+      nextBatch = root;
+    }
+  }
+
+  @Override
+  public boolean hasNext() {
+    return nextBatch != null;
+  }
+
+  /**
+   * Gets the next vector. The user is responsible for freeing its resources.
+   */
+  public VectorSchemaRoot next() {
+    Preconditions.checkArgument(hasNext());
+    VectorSchemaRoot returned = nextBatch;
+    try {
+      load(VectorSchemaRoot.create(rootSchema, allocator));
+    } catch (Exception e) {
+      close();
```

Comment:
I think we should rely on the user of the interface to close it (unless close is idempotent).